src/packet.ml

```diff
@@ -346,42 +385,50 @@ module Tls_crypt = struct
         assert (rsid_len <> 0);
         Cstruct.BE.set_uint64 buf (clear_hdr_len + 1 + acks_len) v)
       hdr.remote_session;
-    (buf, clear_hdr_len + 1 + acks_len + rsid_len)
+    clear_hdr_len + 1 + acks_len + rsid_len
+
+  let encode_control buf (header, sequence_number, payload) =
+    let len = encode_header buf header in
+    Cstruct.BE.set_uint32 buf len sequence_number;
+    Cstruct.blit payload 0 buf (len + 4) (Cstruct.length payload)
 
-  let encode_control op (header, sequence_number, payload) =
-    let hdr_buf, len = encode_header header in
-    let sequence_number_buf, len = (Cstruct.create 4, len + 4) in
-    Cstruct.BE.set_uint32 sequence_number_buf 0 sequence_number;
+  let encode proto (key, (p : [< `Ack of header | `Control of _ ])) =
+    let hdr = header p in
     let len =
-      match op with
-      | Hard_reset_client_v3 ->
-          (* In Hard_reset_client_v3 we don't consider wKc part of the payload *)
-          len
-      | _ -> len + Cstruct.length payload
+      let len_acks = id_len * List.length hdr.ack_sequence_numbers in
+      protocol_len proto + 1 + hdr_len hmac_len + len_acks
+      + (if len_acks = 0 then 0 else 8)
+      +
+      match p with
+      | `Ack _ -> 0
+      | `Control (_, (_, _, payload)) -> 4 + Cstruct.length payload
     in
-    (Cstruct.concat [ hdr_buf; sequence_number_buf; payload ], len)
-
-  let encode proto (key, p) =
-    (* here [len] is the length of the data that is considered part of the packet;
-       for Hard_reset_client_v3 the wKc is appended after the packet. Thus
-       [len] may be shorter than [Cstruct.length payload]. *)
-    let payload, len =
+    let wkc_len =
       match p with
-      | `Ack ack -> encode_header ack
-      | `Control (op, control) -> encode_control op control
+      | `Control (Hard_reset_client_v3, (_, _, wkc)) -> Cstruct.length wkc
+      | _ -> 0
     in
-    let op_buf =
-      let b = Cstruct.create 1 in
-      let op = op_key (operation p) key in
-      Cstruct.set_uint8 b 0 op;
-      b
+    let buf = Cstruct.create len in
```

Comment:
further above, you used create_unsafe -- we may use that here as well?

Reply:
I pushed a commit that uses `create_unsafe`. I chose not to use `create_unsafe` at first because allocation and filling were done in different parts of the code. But it should be safe!

Reply:
:D